File: diffkemp/building/cc_wrapper/cc_wrapper.py
#!/usr/bin/env python
"""
Template for C compiler wrapper used in build stage of DiffKemp

It's written to be either run with RPython to generate a native code
executable, or with Python to modify itself into a Python script.
Because of that it has to run under both Python 2 and Python 3 and some of
Python library functions are redefined here to be RPython-compatible.
"""
# TODO: Decompose simultaneous compiling and linking to get individual .ll
#       files out of it; support other GCC languages than C (Fortran, C++)
from diffkemp.llvm_ir.compiler import get_clang_default_options
import os
import shutil
import sys


wrapper_env_vars = {
    "db_filename": "DIFFKEMP_WRAPPER_DB_FILENAME",
    "clang": "DIFFKEMP_WRAPPER_CLANG",
    "clang_append": "DIFFKEMP_WRAPPER_CLANG_APPEND",
    "clang_drop": "DIFFKEMP_WRAPPER_CLANG_DROP",
    "debug": "DIFFKEMP_WRAPPER_DEBUG",
    "llvm_link": "DIFFKEMP_WRAPPER_LLVM_LINK",
    "llvm_dis": "DIFFKEMP_WRAPPER_LLVM_DIS",
    "no_opt_override": "DIFFKEMP_WRAPPER_NO_OPT_OVERRIDE",
}


def get_cc_wrapper_path(no_native_cc_wrapper=False):
    """
    Find C compiler wrapper. The priority is as follows:
    1. cc_wrapper-c inside build directory
    2. diffkemp-cc-wrapper from PATH
    3. wrapper source file (executable) = this file (if executable)
    4. diffkemp-cc-wrapper.py from PATH (usually executable copy of this file)
    :param no_native_cc_wrapper: Always use wrapper source file
    """
    # Note: this function is ignored by RPython
    from diffkemp.utils import get_simpll_build_dir
    wrapper_exe = os.path.join(os.path.abspath(get_simpll_build_dir()),
                               "cc_wrapper-c")
    if os.path.exists(wrapper_exe) and not no_native_cc_wrapper:
        return wrapper_exe
    elif shutil.which("diffkemp-cc-wrapper") is not None \
            and not no_native_cc_wrapper:
        return "diffkemp-cc-wrapper"
    elif os.access(__file__, os.X_OK):
        # Use wrapper source, i.e. this file
        return __file__
    else:
        return "diffkemp-cc-wrapper.py"


def execl(file, args):
    # Note: execl from os is not RPython-compatible
    if os.path.sep in file:
        os.execv(file, args)
        return
    if "PATH" in os.environ.keys():
        envpath = os.environ["PATH"]
    else:
        envpath = os.defpath
    path = envpath.split(os.pathsep)
    saved_exc = None
    last_exc = None
    for directory in path:
        fullname = os.path.join(directory, file)
        try:
            os.execv(fullname, args)
        except OSError as e:
            last_exc = e
            if saved_exc is None:
                saved_exc = e
    if saved_exc is not None:
        raise saved_exc
    raise last_exc


class CalledProcessError(OSError):
    pass


def check_call(file, args):
    pid = os.fork()
    if pid == 0:
        execl(file, args)
    _, status = os.waitpid(pid, 0)
    if status != 0:
        raise CalledProcessError()


def wrapper(argv):
    # Get arguments from environmental variables
    db_filename = os.environ.get(wrapper_env_vars["db_filename"])
    clang = os.environ.get(wrapper_env_vars["clang"])
    append = os.environ.get(wrapper_env_vars["clang_append"])
    drop = os.environ.get(wrapper_env_vars["clang_drop"])
    debug = os.environ.get(wrapper_env_vars["debug"])
    llvm_link = os.environ.get(wrapper_env_vars["llvm_link"])
    no_opt_override = \
        os.environ.get(wrapper_env_vars["no_opt_override"]) == "1"

    if (db_filename is None or clang is None or append is None or
            drop is None or argv is None or llvm_link is None):
        print("cc_wrapper: fatal error: missing environmental variable")
        return 1
    append = append.split(",")
    drop = drop.split(",")

    db = []
    # Run GCC
    argv[0] = "gcc"
    try:
        check_call("gcc", argv)
    except CalledProcessError:
        print("cc_wrapper: warning: original build command failed")
        return 1

    # Analyze and modify parameters for clang (phase 1)
    clang_argv = []
    old_clang = clang
    linking_with_sources = False
    output_file = None
    linking = "-c" not in argv
    # Check if arguments contains C source file
    contains_source = False
    for index, arg in enumerate(argv):
        if arg in drop:
            continue
        is_object_file = (arg.endswith(".o") or arg.endswith(".lo") or
                          arg.endswith(".ko"))
        is_source_file = arg.endswith(".c")
        contains_source = contains_source or is_source_file
        if index > 1 and argv[index - 1] == "-o":
            if is_object_file and not linking:
                # Compiling to object file: swap .o with .bc
                arg = arg.rsplit(".", 1)[0] + ".bc"
            if not is_object_file and linking:
                # Linking: add a .bcw suffix (LLVM IR whole)
                arg = arg + ".bcw"
            output_file = arg
        elif is_object_file and linking:
            # Input to linking phase: change suffix to .bc
            arg = arg.rsplit(".", 1)[0] + ".bc"
            clang = llvm_link
        elif is_source_file and linking:
            # Mark as linking with sources to detect hybrid mode
            linking_with_sources = True
        clang_argv.append(arg)

    if linking_with_sources and clang == llvm_link:
        # Compile/link mode with object files detected
        # Drop object files and revert to normal compile/link mode
        clang = old_clang
        clang_argv = [arg for arg in clang_argv if not arg.endswith(".bc")]

    # Do not continue if output is not .bc
    # Note: this means that this is neither compilation nor linking
    if (output_file is not None and not output_file.endswith(".bc") and
            not output_file.endswith(".bcw")):
        return 0

    # Do not run clang on conftest files
    if output_file in ["conftest.bc", "conftest.bcw"] or "conftest.c" in argv:
        return 0

    # Not compiling C source file
    if not linking and not contains_source:
        return 0

    # Record file in database
    if output_file is not None:
        prefix = "o:" if clang != llvm_link else "f:"
        db.append(prefix + os.path.join(os.getcwd(), output_file))
    elif not linking:
        # Compiling to default output file
        db.extend(["o:" + os.path.join(os.getcwd(),
                                       arg.rsplit(".", 1)[0] + ".bc")
                   for arg in clang_argv if not arg.endswith(".c")])

    # Analyze and modify parameters for clang (phase 2)
    clang_argv[0] = clang
    if clang != llvm_link:
        # Note: clang uses the last specified optimization level so
        # extending with the default options must be done before
        # extending with the clang_append option.
        clang_argv.extend(get_clang_default_options(
            default_optim=not no_opt_override))
        clang_argv.extend(append)
        # TODO: allow compiling into binary IR
    else:
        # Keep only arguments with input files (and llvm-link itself)
        clang_argv = [arg for arg in clang_argv if arg == clang or
                      arg.endswith(".bc") or arg.endswith(".bcw") or
                      arg == "-o"]
        # Remove non-existent files
        # Note: these might have been e.g. generated from assembly
        new_clang_argv = [clang_argv[0], "-S"]
        output_file_flag = False
        for arg in clang_argv[1:]:
            if output_file_flag or os.path.exists(arg) or arg == "-o":
                new_clang_argv.append(arg)
                if output_file_flag:
                    output_file_flag = False
            if arg == "-o":
                output_file_flag = True
        clang_argv = new_clang_argv

    # Run clang
    try:
        if debug:
            print("Wrapper calling: " + " ".join(clang_argv))
        check_call(clang, clang_argv)
    except CalledProcessError:
        print("cc_wrapper: warning: clang failed")
        print(clang_argv)
        # clang error is non-fatal
        return 0

    # Update database
    with open(db_filename, "a") as db_file:
        # TODO: RPython-compatible file locking
        for entry in db:
            if not os.path.exists(entry.split(":")[1]):
                continue
            db_file.write(entry + "\n")

    return 0


def target(driver, args):
    return wrapper


# Enable running directly under Python interpreter
if __name__ == "__main__":
    wrapper(sys.argv)

File: diffkemp/building/cc_wrapper/cc_wrapper.cpp
/**
 * Diffkemp Compiler Wrapper
 *
 * internal args:
 * dbf   : <string> Path to the output snapshot database (db_filename).
 * clang : <string> Path to the clang binary.
 * cla   : vector<string> "Clang Append" - extra flags to pass to clang.
 * cld   : unordered_set<string> "Clang Drop" - flags to remove from the clang
 * invocation.
 * debug : <bool> Enable debug logging if present.
 * llink : <string> Path to the llvm-link binary.
 * lldis : <string> Path to the llvm-dis.
 * binary.
 * noo   : <bool> "No Opt Override" - do not override optimization levels.
 *
 * Arguments following these internal args are stored in `cFlags` and passed
 * to the underlying compiler (gcc) and clang.
 */

#include "cc_wrapper_utils.h"
#include <algorithm>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <sys/wait.h>
#include <unistd.h>
#include <utility>
#include <vector>

namespace fs = std::filesystem;

int checkCall(const std::string &cmd, const std::vector<std::string> &args) {
    std::vector<char *> cArgs;
    // +2 for the command itself and the nullptr terminator
    cArgs.reserve(args.size() + 2);

    cArgs.push_back(const_cast<char *>(cmd.c_str()));

    for (const auto &arg : args) {
        cArgs.push_back(const_cast<char *>(arg.c_str()));
    }

    cArgs.push_back(nullptr);

    pid_t pid = fork();

    if (pid == -1) {
        std::cerr << "fork failed" << "\n";
        return -1;
    }

    if (pid == 0) {
        execvp(cmd.c_str(), cArgs.data());
        std::cerr << "execvp failed for: " << cmd << "\n";
        // command not found
        _exit(127);
    }

    int status;

    if (waitpid(pid, &status, 0) == -1) {
        return -1;
    }
    if (!WIFEXITED(status)) {
        return EXIT_FAILURE;
    }

    return WEXITSTATUS(status);
}

int wrapper(int argc, char *argv[]) {
    WrapperArgs wrapperArgs = WrapperArgs(argc, argv);

    // Run GCC
    if (const int status = checkCall("gcc", wrapperArgs.cFlags); status != 0) {
        std::cerr << "cc_wrapper warning: original build command failed with "
                     "code "
                  << status << "\n";
        return status;
    }

    // Analyze and modify parameters for clang (phase 1)
    std::vector<std::string> clangArgv;
    std::string clangBin = wrapperArgs.clang;
    bool linkingWithSources = false;
    std::string outputFile;
    bool linking = !isInVector("-c", wrapperArgs.cFlags);
    // Check if arguments contains C source files
    bool containsSource = false;
    for (std::size_t i = 0; i < wrapperArgs.cFlags.size(); ++i) {
        std::string arg = wrapperArgs.cFlags[i];
        if (wrapperArgs.clangDrop.find(arg) != wrapperArgs.clangDrop.end()) {
            continue;
        }

        bool isObjFile = (endsWith(arg, ".o") || endsWith(arg, ".lo")
                          || endsWith(arg, ".ko"));
        bool isSourceFile = endsWith(arg, ".c");
        containsSource = containsSource || isSourceFile;
        if (i > 0 && wrapperArgs.cFlags[i - 1] == "-o") {
            if (isObjFile && !linking) {
                // Compiling to object file: swap .o with .ll
                arg = arg.substr(0, arg.rfind('.')) + ".ll";
            }
            if (!isObjFile && linking) {
                // Linking: add a .llw suffix (LLVM IR whole)
                arg = arg + ".llw";
            }
            outputFile = arg;
        } else if (isObjFile && linking) {
            // Input to linking phase: change suffix to .ll
            arg = arg.substr(0, arg.rfind('.')) + ".ll";
            clangBin = wrapperArgs.llvmLink;
        } else if (isSourceFile && linking) {
            // Mark as linking with sources to detect hybrid mode
            linkingWithSources = true;
        }
        clangArgv.push_back(arg);
    }

    if (linkingWithSources && clangBin == wrapperArgs.llvmLink) {
        // Compile/link mode with object files detected
        // Drop object files and revert to normal compiler/link mode
        clangBin = wrapperArgs.clang;
        std::vector<std::string> tmp;
        std::copy_if(clangArgv.begin(),
                     clangArgv.end(),
                     std::back_inserter(tmp),
                     [](const auto &arg) { return !endsWith(arg, ".ll"); });
        clangArgv = std::move(tmp);
    }

    // Do not continue if output is not .ll or .llw
    // Note: this means that this is netiher compilation nor linking
    if (outputFile.empty()
        || (!endsWith(outputFile, ".ll") && !endsWith(outputFile, ".llw"))) {
        return 0;
    }

    // Do not run clang on conftest files
    if (outputFile == "conftest.ll" || outputFile == "conftest.llw"
        || isInVector("conftest.c", wrapperArgs.cFlags)) {
        return 0;
    }

    // Not compiling C source file
    if (!linking && !containsSource) {
        return 0;
    }

    // Record file in database
    std::vector<std::string> dbEntries;
    if (!outputFile.empty()) {
        std::string prefix = (clangBin != wrapperArgs.llvmLink) ? "o:" : "f:";
        dbEntries.push_back(prefix
                            + (fs::current_path() / outputFile).string());
    } else if (!linking) {
        // Compiling to default output file
        for (const std::string &arg : clangArgv) {
            if (!endsWith(arg, ".c")) {
                fs::path p(arg);
                p.replace_extension(".ll");
                dbEntries.push_back("o:" + (fs::current_path() / p).string());
            }
        }
    }

    // Analyze and modify parameters for clang (phase 2)
    if (clangBin != wrapperArgs.llvmLink) {
        // Note: clang uses the last specified optimization level so
        // extending with the default options must be done before
        // extending with the clang_append option.
        auto defaults = getClangDefaultOptions(!wrapperArgs.noOptOverride);
        clangArgv.insert(clangArgv.end(), defaults.begin(), defaults.end());
        clangArgv.insert(clangArgv.end(),
                         wrapperArgs.clangAppend.begin(),
                         wrapperArgs.clangAppend.end());
        // TODO: allow compiling into binary IR
    } else {
        // Keep only arguments with input files (and llvm-link itself)
        std::vector<std::string> tempArgs;
        for (const auto &arg : clangArgv) {
            if (endsWith(arg, ".ll") || endsWith(arg, ".llw") || arg == "-o") {
                tempArgs.push_back(arg);
            }
        }

        // Remove non-existing files
        // Note: these might have been e.g. generated from assembly
        std::vector<std::string> newArgs = {"-S"};
        bool outputFlag = false;
        for (const auto &arg : tempArgs) {
            if (outputFlag || fs::exists(arg) || arg == "-o") {
                newArgs.push_back(arg);
            }
            outputFlag = arg == "-o";
        }
        clangArgv = std::move(newArgs);
    }

    if (wrapperArgs.debug) {
        std::cout << "Wrapper calling:";
        for (const auto &arg : clangArgv) {
            std::cout << arg;
        }
        std::cout << "\n";
    }

    // Run clang
    if (const int status = checkCall(clangBin, clangArgv); status != 0) {
        if (wrapperArgs.debug) {
            std::cerr << "cc_wrapper warning: clang failed with code " << status
                      << "\n";
        }
        // clang error is non-fatal
        return EXIT_SUCCESS;
    }

    // Update the database file
    std::ofstream db(wrapperArgs.dbFilename, std::ios::app);
    if (db) {
        for (const auto &entry : dbEntries) {
            std::string path = entry.substr(2);
            if (fs::exists(path)) {
                db << entry << "\n";
            }
        }
    } else {
        std::cerr << "cc_wrapper warning: cannot open DB file for append:"
                  << wrapperArgs.dbFilename << "\n";
    }

    return EXIT_SUCCESS;
}

int main(int argc, char *argv[]) { return wrapper(argc, argv); }
